### PR TITLE
Move sse compiler options to PCL_COMPILE_OPTIONS

### LINKED ISSUE #1725: ABI problem - PCL_ENABLE_SSE changes alignment of class members
I noticed in a test program on Ubuntu 16.04, gcc version 5.4.0 20160609 with pcl 1.8 (HEAD version) that a simple CropBox filter generated completely unpredictable results. It took me some hours to track this problem down:

I had built PCL with the default option PCL_ENABLE_SSE but my application was missing the SSE build flags, e.g. specifically `-march=native -msse4.2 -mfpmath=sse`. I simply used `FIND_PACKAGE(PCL 1.8 REQUIRED)` but that currently does not set/configure any SSE related options.

The reason for CropBox behaving unpredictable was uninitialized memory access. The PCL shared library and the calling application generated a slightly different class layout.

I added two debug outputs to see sizes and pointer-locations .. one in the ctor of CropBox in `pcl/filters/crop_box.h` (seen by my test app) and the other in `CropBox::applyFilter` in `crop_box.hpp` (which gets compiled into `libpcl_filters.so`). 

Here is the output that I got:
### Inside CropBox ctor:

```
this:140720843586960

Size infos:
sizeof(CropBox):240
sizeof(FilterIndices):104
sizeof(Filter):104
sizeof(PCLBase):48
sizeof(Affine3f):64
sizeof(Eigen::Vector4f):16
sizeof(Eigen::Vector3f):12

pointer values:
&min_pt_: 140720843587072
&max_pt_: 140720843587088
&translation_: 140720843587104
&rotation_: 140720843587116
&transform_: 140720843587136
```
### Inside applyFilter:

```
this:140720843586960

Sizes:
sizeof(CropBox):256                 (* different!)
sizeof(FilterIndices):104
sizeof(Filter):104
sizeof(PCLBase):48
sizeof(Affine3f):64
sizeof(transform_):64
sizeof(Eigen::Vector4f):16
sizeof(Eigen::Vector3f):12

Member addresses:
&min_pt_: 140720843587072
&max_pt_: 140720843587088
&translation_: 140720843587104
&rotation_: 140720843587116
&transform_: 140720843587152    (* different!)
```

It seems to be very important to specify 'compatible' compiler options for a PCL client-application. 
I wonder whether this problem is new in Unbuntu 16.04? Otherwise I would expect that a lot of people would have seen problems with the `CropBox` filter.

As a solution maybe `PCLConfig.cmake` could set a variable which contains the SSE settings that were used during the PCL build (e.g. `${SSE_FLAGS}`)?

Potentially related:
- https://github.com/PointCloudLibrary/pcl/issues/1623
- http://www.pcl-users.org/cropbox-problems-td4019336.html

### LINKED ISSUE #2013: Simply creating a GeneralizedIterativeClosestPoint6D (gicp) object causes segmentation fault
## Your Environment
* Operating System and version: Ubuntu 16.04
* Compiler: GCC 5.4
* PCL Version: Latest

## Minimal code to Reproduce

CMakeLists

``````
cmake_minimum_required(VERSION 3.7)
project(Example)

set(PROJECT_SRCS ${PROJECT_SOURCE_DIR}/main.cpp )

find_package( PCL REQUIRED )

add_executable(${PROJECT_NAME} ${PROJECT_SRCS})

target_include_directories(${PROJECT_NAME} SYSTEM PUBLIC                          
 ${PCL_INCLUDE_DIRS})

target_link_libraries(${PROJECT_NAME} ${PCL_LIBRARIES})
``````

main.cpp

``````
#include "pcl/registration/gicp6d.h"

int main(){
  pcl::GeneralizedIterativeClosestPoint6D gicp;

  return 0;
}
``````

## Current Behavior
``````
ubuntu@XXX.XXX.XX.XX:~/test/build$ ./Example 
Segmentation fault (core dumped)
``````

## Expected Behavior
``````
ubuntu@XXX.XXX.XX.XX:~/test/build$ ./Example 
ubuntu@XXX.XXX.XX.XX:~/test/build$
``````

## Context
This problem is preventing me from using the GICP function, which I would like to improve my result by replacing the traditional ICP with GICP.

@MichaelKorn Thank you for adding GICP into PCL, could you help me solve this problem?

### PR DESCRIPTION
**Pull request by SergioRAgostinho**
This PR ensures the SSE compiler options are passed properly downstream. After reading the descriptions for 

https://cmake.org/cmake/help/v3.0/command/target_compile_definitions.html#command:target_compile_definitions
https://cmake.org/cmake/help/v3.0/command/target_compile_options.html?highlight=target_compile_options

I opted to create the new var PCL_COMPILE_OPTIONS that should be used by downstream projects like this
```cmake
cmake_minimum_required(VERSION 3.0.0 FATAL_ERROR)
project(my_awesome_project VERSION 0.1.0 LANGUAGES CXX)

find_package(PCL 1.8.1.99 REQUIRED)

add_executable (${PROJECT_NAME} "${PROJECT_NAME}.cpp")
target_include_directories(${PROJECT_NAME} PUBLIC ${PCL_INCLUDE_DIRS})
target_link_libraries (${PROJECT_NAME} ${PCL_LIBRARIES})
target_compile_definitions(${PROJECT_NAME} PUBLIC ${PCL_DEFINITIONS})
target_compile_options(${PROJECT_NAME} PUBLIC ${PCL_COMPILE_OPTIONS})
```

Fixes #2013, fixes #1725

### CODE REVIEW

**Comment by SergioRAgostinho** [comment]
`PCL_COMPILE_OPTIONS` or simply `PCL_OPTIONS`?

**Comment by jasjuang** [comment]
@SergioRAgostinho An even more modern cmake approach is `target_link_libraries` to "library targets" instead of the "plain library".  More explanation here https://cmake.org/cmake/help/v3.9/command/target_link_libraries.html. The targets will contain libraries, includes, and definitions. So in the downstream cmake it will look something like

```
cmake_minimum_required(VERSION 3.9 FATAL_ERROR)
project(my_awesome_project LANGUAGES CXX)

find_package(PCL 1.8.1.99 REQUIRED)

add_executable (${PROJECT_NAME} "${PROJECT_NAME}.cpp")

target_link_libraries (${PROJECT_NAME} pcl::pcl)
```

Thanks a lot for attempting to fix #2013

**Comment by SergioRAgostinho** [comment]
Agreed but that's a massive rework into the whole CMake system. One step at the time. 

Also the SSE flags are compiler options, not exactly preprocessor definitions. I'm not sure those are automatically propagated.

**Comment by jasjuang** [comment]
@SergioRAgostinho My understanding is if you are building a library in the upstream and exporting it as a target for the downstream to target_link, no matter you are doing `target_compile_definitions` or 
`target_compile_options` in the upstream, the information will be embedded in the target and get passed to the downstream.

**Comment by SergioRAgostinho** [comment]
So I have the impression that works pretty great if all your targets are from within the same project, or being built in the same system, but when you want to deploy your library on some external system things are a little different no?

In theory you're shipping only your headers, libs, a cmake config with info about which components were built and which external dependencies used/enabled.

Do you have an example I can check of a situation where you linked against a third party which implemented this system properly?

**Comment by taketwo** [comment]
> Agreed but that's a massive rework into the whole CMake system. One step at the time.

Actually a very basic implementation of this is possible without any rework, just changing a bunch of lines in `PCLConfig`. A can give it a try.

**Comment by jasjuang** [comment]
@SergioRAgostinho The smallest 3rd party library I know that does this properly is [glog](https://github.com/google/glog). Only `target_link_libraries (${PROJECT_NAME} glog::glog)` is needed for the downstream. Another even smaller 3rd party library that also does this almost right is [tinyxml2](https://github.com/leethomason/tinyxml2/), `target_link_libraries (${PROJECT_NAME} tinyxml2)` will also just work, but they didn't add the namespace (`tinyxml2::tinyxml2`) which is considered as good practice.

CMake actually allows you export targets differently for internal and external stuff. One classical example is the DLL macros for windows. You can do something like this

```
if(MSVC)
  set(DLLIMPORT "__declspec(dllimport)")
  set(DLLEXPORT "__declspec(dllexport)")
endif()

target_compile_definitions(${PROJECT_NAME} INTERFACE
                          WINDOWS_DLL=${DLLIMPORT})

target_compile_definitions(${PROJECT_NAME} PRIVATE
                          WINDOWS_DLL=${DLLEXPORT})
```

So externally the macro is dllimport to them, internally it's dllexport.

**Comment by SergioRAgostinho** [comment]
These exports then...

```cmake
install (TARGETS glog
  EXPORT glog-targets
  RUNTIME DESTINATION ${_glog_CMake_BINDIR}
  PUBLIC_HEADER DESTINATION ${_glog_CMake_INCLUDE_DIR}/glog
  LIBRARY DESTINATION ${_glog_CMake_LIBDIR}
  ARCHIVE DESTINATION ${_glog_CMake_LIBDIR})


export (TARGETS glog NAMESPACE glog:: FILE glog-targets.cmake)
export (PACKAGE glog)

install (FILES
  ${CMAKE_CURRENT_BINARY_DIR}/glog-config.cmake
  ${CMAKE_CURRENT_BINARY_DIR}/glog-config-version.cmake
  DESTINATION ${_glog_CMake_INSTALLDIR})

install (EXPORT glog-targets NAMESPACE glog:: DESTINATION
${_glog_CMake_INSTALLDIR})
```

**Comment by taketwo** [comment]
Yeah, exporting the targets is "the right way", and this will require serious rework.
Temporary hacky solution is to manually create imported targets inside PCLConfig. Here is my try on it:

```diff
diff --git a/PCLConfig.cmake.in b/PCLConfig.cmake.in
index 7aed0a7..a7e3dea 100644
--- a/PCLConfig.cmake.in
+++ b/PCLConfig.cmake.in
@@ -536,9 +536,9 @@ foreach(component ${PCL_TO_FIND_COMPONENTS})
       else(PCL_${COMPONENT}_LIBRARY_DEBUG)
         list(APPEND PCL_${COMPONENT}_LIBRARIES ${PCL_${COMPONENT}_LIBRARY})
       endif(PCL_${COMPONENT}_LIBRARY_DEBUG)
-      list(APPEND PCL_LIBRARIES ${PCL_${COMPONENT}_LIBRARIES})
+      list(APPEND PCL_LIBRARIES ${pcl_component})
       mark_as_advanced(PCL_${COMPONENT}_LIBRARY PCL_${COMPONENT}_LIBRARY_DEBUG)
-    endif(_is_header_only EQUAL -1)    
+    endif()
     # Append internal dependencies
     foreach(int_dep ${pcl_${component}_int_dep})
       string(TOUPPER "${int_dep}" INT_DEP)
@@ -549,6 +549,30 @@ foreach(component ${PCL_TO_FIND_COMPONENTS})
         endif(PCL_${INT_DEP}_LIBRARIES) 
       endif(PCL_${INT_DEP}_FOUND)
     endforeach(int_dep)
+    if(_is_header_only EQUAL -1)
+      add_library(${pcl_component} @PCL_LIB_TYPE@ IMPORTED GLOBAL)
+      if(PCL_${COMPONENT}_LIBRARY_DEBUG)
+        set_target_properties(${pcl_component}
+          PROPERTIES
+            IMPORTED_CONFIGURATIONS "RELEASE;DEBUG"
+            IMPORTED_LOCATION_RELEASE "${PCL_${COMPONENT}_LIBRARY}"
+            IMPORTED_LOCATION_DEBUG "${PCL_${COMPONENT}_LIBRARY_DEBUG}"
+        )
+      else()
+        set_target_properties(${pcl_component}
+          PROPERTIES
+            IMPORTED_LOCATION "${PCL_${COMPONENT}_LIBRARY}"
+        )
+      endif()
+      set_target_properties(${pcl_component}
+        PROPERTIES
+          COMPILE_DEFINITIONS "${PCL_${COMPONENT}_DEFINITIONS}"
+          COMPILE_OPTIONS ""
+          INTERFACE_INCLUDE_DIRECTORIES "${PCL_${COMPONENT}_INCLUDE_DIRS}"
+          INTERFACE_LINK_LIBRARIES "${PCL_${COMPONENT}_LIBRARIES}"
+      )
+      set(PCL_${COMPONENT}_LIBRARIES ${pcl_component})
+    endif()
   endif(PCL_${COMPONENT}_FOUND)
 endforeach(component)
```

Haven't tested it yet, my laptop needs ages to build PCL.

**Comment by SergioRAgostinho** [comment]
Applied the patch
```
CMake Error at /home/sergio/Development/3rdparty/pcl/install/share/pcl-1.8/PCLConfig.cmake:609 (set_target_properties):
  Property INTERFACE_LINK_LIBRARIES may not contain link-type keyword
  "optimized".  The INTERFACE_LINK_LIBRARIES property may contain
  configuration-sensitive generator-expressions which may be used to specify
  per-configuration rules.
Call Stack (most recent call first):
  CMakeLists.txt:4 (find_package)
```

**Comment by taketwo** [comment]
Updated the diff. Should be fixed now; works on my machine. How this will behave on Windows I am not sure though.

**Comment by SergioRAgostinho** [comment]
:+1: Looking good. The SSE compiler definitions are not being propagated though.
```
/usr/bin/c++   -Dqh_QHpointer -isystem /usr/include/vtk-5.10 -isystem /usr/include/eigen3 -isystem /home/sergio/Development/3rdparty/pcl/install/include/pcl-1.8 -isystem /usr/include/ni -isystem /usr/include/openni2  -Wno-deprecated -g   -o CMakeFiles/gicp_double_free.dir/gicp_double_free.cpp.o -c /home/sergio/Development/pcl_test_cases/gicp_double_free/gicp_double_free.cpp
```

**Comment by taketwo** [comment]
This is how I set definitions/options:

```diff
+          COMPILE_DEFINITIONS "${PCL_${COMPONENT}_DEFINITIONS}"
+          COMPILE_OPTIONS ""
```

I suspect component definitions are actually empty. So just apply `PCL_DEFINITIONS`. (Updated the diff again).

**Comment by SergioRAgostinho** [comment]
I tried to set the compiler options in `COMPILE_OPTIONS` by passing the `PCL_COMPILE_OPTIONS` list but it is not cascading down.

**Comment by taketwo** [comment]
You are applying this diff on top of your pull request, right?

**Comment by SergioRAgostinho** [comment]
Yup

**Comment by taketwo** [comment]
One more try, prefixed property names with `INTERFACE_`

**Comment by SergioRAgostinho** [comment]
We're getting somewhere but we need to extract that -D prefix

[ 50%] Building CXX object CMakeFiles/gicp_double_free.dir/gicp_double_free.cpp.o
/usr/bin/c++   -D-DDISABLE_DAVIDSDK -D-DDISABLE_DSSDK -D-DDISABLE_ENSENSO -D-DDISABLE_LIBUSB_1_0 -D-DDISABLE_PCAP -D-DDISABLE_PNG -Dqh_QHpointer -isystem /usr/include/vtk-5.10 -isystem /usr/include/eigen3 -isystem /home/sergio/Development/3rdparty/pcl/install/include/pcl-1.8 -isystem /usr/include/ni -isystem /usr/include/openni2  -Wno-deprecated -g   -march=native -o CMakeFiles/gicp_double_free.dir/gicp_double_free.cpp.o -c /home/sergio/Development/pcl_test_cases/gicp_double_free/gicp_double_free.cpp

**Comment by taketwo** [comment]
Also, somebody [suggested](https://stackoverflow.com/a/24303754/1525865) that options should be a list, not a string.

**Comment by SergioRAgostinho** [comment]
Yep. I confirmed that my sse flags are now passed the correctly but the definitions are messed up because they're being stored with "-D" prefix.

**Comment by taketwo** [comment]
That's a simple regex, wait a sec

**Comment by SergioRAgostinho** [comment]
We'll also need to define the pcl namespace to avoid target name collision with whoever works with our project downstream.

**Comment by taketwo** [comment]
> We'll also need to define the pcl namespace to avoid target name collision with whoever works with our project downstream.

Right now targets are prefixed, e.g. `pcl_common`, so collisions are very unlikely. If anyone creates such targets in his project, then it's... his fault :) But yeah, "the right way" will also have namespacing.

**Comment by taketwo** [comment]
Last try for today. (The filtering code will need to be factored into utility function later.)

**Comment by SergioRAgostinho** [comment]
:+1: Gonna append your diff as your commit.

File: cmake/pcl_find_sse.cmake [hunk 1]
```diff
@@ -169,27 +169,27 @@ macro(PCL_CHECK_FOR_SSE)
...
             # platforms.
-            set(SSE_FLAGS "-ffloat-store")
+            list(APPEND SSE_FLAGS "-ffloat-store")
         endif()
     elseif(MSVC AND NOT CMAKE_CL_64)
         if(HAVE_SSE2_EXTENSIONS)
-            SET(SSE_FLAGS "${SSE_FLAGS} /arch:SSE2")
+            list(APPEND SSE_FLAGS "/arch:SSE2")
         elseif(HAVE_SSE_EXTENSIONS)
-            SET(SSE_FLAGS "${SSE_FLAGS} /arch:SSE")
```
Comment (taketwo) [inline cmake/pcl_find_sse.cmake]: Is this on purpose that `SSE` became `SSE3`?

**Comment by taketwo** [comment]
I think it's essential that these changes are tested on Windows before merging. @UnaNancyOwen can you give it a try? Important is to test if after building/installing PCL dependent projects can successfully link against. Further, switching between Debug/Release builds should be possible.

**Comment by UnaNancyOwen** [comment]
@taketwo Yes, I'm going to try this tonight.

File: cmake/pcl_find_sse.cmake (hunk 1 above)
Comment (SergioRAgostinho) [reply cmake/pcl_find_sse.cmake]: Bad copy paste.

**Comment by UnaNancyOwen** [comment]
It doesn't seem to be working correctly.
PCL libraries is correctly displayed in CMake-GUI.
But, <code>PCL_LIBRARIES</code> doesn't correctly have the PCL libraries when configure CMake.
Therefor, It is not set correctly to Visual C++ project that generated by CMake.

For example
* PCL_COMMON_LIBRARY : <code>C:/Program Files/PCL/lib/pcl_common_release.lib</code>
* PCL_COMMON_LIBRARY_DEBUG : <code>C:/Program Files/PCL/lib/pcl_common_debug.lib</code>
* PCL_LIBRARIES (part of results) : <code>pcl_common</code>

Correctly results of PCL_LIBRARIES is <code>optimized;C:/Program Files/PCL/lib/pcl_common_release.lib;debug;C:/Program Files/PCL/lib/pcl_common_debug.lib;</code>.

cmake_pcl

```cmake
cmake_minimum_required( VERSION 2.8 )

project( solution )
add_executable( project main.cpp )

find_package( PCL 1.8.1.99 REQUIRED )

if( PCL_FOUND )
  include_directories( ${PCL_INCLUDE_DIRS} )
  message(STATUS "PCL_INCLUDE_DIRS    : ${PCL_INCLUDE_DIRS}")

  add_definitions( ${PCL_DEFINITIONS} )
  message(STATUS "PCL_DEFINITIONS     : ${PCL_DEFINITIONS}")
  
  add_compile_options( ${PCL_COMPILE_OPTIONS} )
  message(STATUS "PCL_COMPILE_OPTIONS : ${PCL_COMPILE_OPTIONS}")

  link_directories( ${PCL_LIBRARY_DIRS} )
  message(STATUS "PCL_LIBRARY_DIRS    : ${PCL_LIBRARY_DIRS}")

  target_link_libraries( project ${PCL_LIBRARIES} )
  message(STATUS "PCL_LIBRARIES       : ${PCL_LIBRARIES}")
endif()
```

In this case, It is correct that <code>PCL_COMPILE_OPTIONS</code> is empty because it is  configuration of Visual C ++ Win64 (x64). It will include <code>/arch:SSE /arch:SSE2</code> when x86 configuration.

BTW : 
PCLConfig will it become possible to specify it like pcl::pcl?
I'm not familiar with new method. How to use it in CMakeLists?
```patch
- # old method
- include_directories( ${PCL_INCLUDE_DIRS} )
- add_definitions( ${PCL_DEFINITIONS} )
- add_compile_options( ${PCL_COMPILE_OPTIONS} )
- target_link_libraries( project ${PCL_LIBRARIES} )

+ # new method
+ target_link_libraries( project pcl::pcl )
```

**Comment by SergioRAgostinho** [comment]
Like this 

```cmake
target_link_libraries( project ${PCL_LIBRARIES})
```

**Comment by UnaNancyOwen** [comment]
If you need more tests on Windows, I will do it and report.

**Comment by SergioRAgostinho** [comment]
Is it working when you specify target_link_libraries properly?

**Comment by UnaNancyOwen** [comment]
Additional Dependencies is set as follows.
Correctly settings of Additional Dependencies is "C:\Program Files\PCL\lib\pcl_common_release.lib". 
vc_pcl

In addition, <code>PCL_LIBRARY_DIRS</code> was not set to Additional Library Directories of Visual C++ project that generated by CMake. It is empty.
Correctly settings of Additional Library Directories is "C:\Program Files\PCL\lib", but it is empty.

It able to build and works correctly when manually fix these settings.

**Comment by taketwo** [comment]
Thanks for testing.

> PCL_COMMON_LIBRARY : C:/Program Files/PCL/lib/pcl_common_release.lib
> PCL_COMMON_LIBRARY_DEBUG : C:/Program Files/PCL/lib/pcl_common_debug.lib
> PCL_LIBRARIES (part of results) : pcl_common;
> Correctly results of PCL_LIBRARIES is optimized;C:/Program Files/PCL/lib/pcl_common_release.lib;debug;C:/Program Files/PCL/lib/pcl_common_debug.lib;.

This is as expected. Instead of explicitly listing the paths to release/debug libraries, we create "imported" targets, e.g. `pcl_common`. When you write `target_link_libraries(your_program pcl_common)`, CMake figures out itself the path to appropriate build configuration and links against it. It also automatically test compiler options and definitions on `your_program`. Therefore `${PCL_LIBRARIES}` now contains only the list of "imported" targets (i.e. PCL modules).

Ok, I'll dust off my Windows laptop and give it a try as well.

**Comment by taketwo** [comment]
Unfortunately won't have time to play with it on Windows any any foreseeable future.

**Comment by jasjuang** [comment]
IMO this PR is still useful to merge since it fixes the problem on linux

**Comment by taketwo** [comment]
Well yes, for now we can merge it without "my" part (imported targets). Then the correct way to link against PCL will be:

```cmake
target_include_directories(${PROJECT_NAME} PUBLIC ${PCL_INCLUDE_DIRS})
target_link_libraries (${PROJECT_NAME} ${PCL_LIBRARIES})
target_compile_definitions(${PROJECT_NAME} PUBLIC ${PCL_DEFINITIONS})
target_compile_options(${PROJECT_NAME} PUBLIC ${PCL_COMPILE_OPTIONS})
```

This is ugly and nobody will do this. But at least it will fix the problem for those who know and do.

**Comment by SergioRAgostinho** [comment]
Sergey does have a point.

@UnaNancyOwen we might need a final push from your side on this one, to make things work on windows.

**Comment by UnaNancyOwen** [comment]
I confirmed that it works correctly without commit https://github.com/PointCloudLibrary/pcl/pull/2100/commits/b7fce9770558b7fd4d6ee184389ed968488fda03 on Windows.

I'm not familiar with imported targets. 
So, @jasjuang Could you check generate project using imported targets on Windows instead of me.

**Comment by jasjuang** [comment]
@UnaNancyOwen I am still stuck on https://github.com/Microsoft/vcpkg/issues/1968, and I never build pcl on windows successfully without vcpkg. I will come back to this after I resolve https://github.com/Microsoft/vcpkg/issues/1968. The problem I am having is building vtk with python feature package is failing for me. I will post the error message over there.

**Comment by jasjuang** [comment]
@SergioRAgostinho @taketwo @UnaNancyOwen Since I resolved https://github.com/Microsoft/vcpkg/issues/1968 I am back to test this out. The way I tested this is checkout this PR and copy the entire source code into the vcpkg `buildtrees\pcl\src\pcl-pcl-1.8.1` folder and install pcl by `vcpkg install pcl`. It compiles successfully, but when I try to use it with a minimal downstream code like the below

CMakeLists.txt
```
cmake_minimum_required(VERSION 3.9)

project(Example)

set(PROJECT_SRCS ${PROJECT_SOURCE_DIR}/mycpp.cpp) 

find_package(PCL REQUIRED)

add_executable(${PROJECT_NAME} ${PROJECT_SRCS})

target_link_libraries(${PROJECT_NAME} ${PCL_LIBRARIES})
```

mycpp.cpp is as simple as 
```
int main(){
return 0;
}
```

The below error shows when I am configuring cmake for my downstream

```
CMake Error at C:/dev/vcpkg/installed/x64-windows-dev/share/pcl/PCLConfig.cmake:46 (message):
  PCL can not be found on this machine
Call Stack (most recent call first):
  C:/dev/vcpkg/installed/x64-windows-dev/share/pcl/PCLConfig.cmake:429 (pcl_report_not_found)
  C:/dev/vcpkg/scripts/buildsystems/vcpkg.cmake:248 (_find_package)
  CMakeLists.txt:8 (find_package)
```

**Comment by taketwo** [comment]
Are you sure this error is caused by this PR? CMake is trying to locate `pcl_config.h` file and fails, but this behavior was not altered by the commits here.

**Comment by andersgb** [comment]
Tested PCL compiled with this PR, and with the suggested CMakeLists.txt setup in my downstream project on Ubuntu 17.10. But I am unable to compile at all, due to:
```
CMake Error in CMakeLists.txt:
  Imported target "pcl_io" includes non-existent path

    "/usr/include/x86_64-linux-gnu/freetype2"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```
The bogus directory is a part of `${PCL_IO_INCLUDE_DIRS}` with and without this PR, but compiles without this PR.

Is this a potential regression? Does CMake have stricter rules on `INTERFACE_INCLUDE_DIRECTORIES` pointing to existing directories?

**Comment by taketwo** [comment]
@andersgb thanks for reporting. This bogus directory clearly comes from VTK. I push a commit that filters the list of includes. Please give a try.

**Comment by taketwo** [comment]
@SergioRAgostinho I think I've just caught a subtle bug in this PR which would only manifest later on when 1.9.0 is packaged and shipped to users via `apt-get`. Pay attention to [this if statement](https://github.com/PointCloudLibrary/pcl/blob/master/CMakeLists.txt#L111). Our `SSE_FLAGS` are appended to `CMAKE_CXX_FLAGS` only if they are empty. When they are **not empty** (which is the case for debian packaging builds), SSE flags are not appended. Yet they are still unconditionally exported to our CMake config. As a result, we have a problem which is sort of the reverse of what we are solving here: PCL deb package is built without `march=native` (for obvious reason), but configure script adds this flag to dependee projects.

**Comment by andersgb** [comment]
@taketwo great, can confirm compilation works now.

**Comment by svenevs** [comment]
I was going to open a separate issue, but this is already in motion and seemed like the right place to put it: **clang support is broken**.

1. The "initial" clang setup [is commented out](https://github.com/SergioRAgostinho/pcl/blob/7b1be8f2bac84d9614a59c558c22d577f55be7af/cmake/pcl_find_sse.cmake#L7-L12).  It should just be

    ```cmake
    if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
      list(APPEND SSE_FLAGS "-march=native")
    endif()
    ```

2. There is no such thing as `CMAKE_COMPILER_IS_CLANG`.  Since it is checked a lot, you could use the above and set something first

    ```cmake
    if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
      set(COMPILER_IS_CLANG TRUE)
    else()
      set(COMPILER_IS_CLANG FALSE)
    endif()
    ```

3. In one instance (only) there is a check:

    ```cmake
    if(GCC_VERSION_STRING VERSION_GREATER 4.2 AND NOT APPLE AND NOT CMAKE_CROSSCOMPILING)
    ```

    Shouldn't it be globally disabled if `CMAKE_CROSSCOMPILING` is true for **all** compilers?  So at the very beginning

    ```cmake
    if (CMAKE_CROSSCOMPILING)
        set(SSE_FLAGS "")
        set(SSE_DEFINITIONS "")
        set(SSE_FLAGS_STR "")
    else()
        # indent everything here
    endif()
    ```

**Comment by SergioRAgostinho** [comment]
1. 
https://github.com/PointCloudLibrary/pcl/blob/fc88d45bcc4f49904bcff1b4359f1310ab516292/CMakeLists.txt#L82-L91

2. is this related to this PR or currently (on master) Clang support is broken?
3. is it Clang or AppleClang?

**Comment by svenevs** [comment]
1. I see, I just snagged the file locally, setting something like `CMAKE_XXX` is unconventional.
2. Related to this PR only in that I saw you were modifying SSE stuff, the same code is commented out here and on master.
3. I'm working with both.  So in light of (1), using `STREQUAL` rather than `MATCHES` means AppleClang is skipped.

So I guess this is mostly unrelated, sorry for adding noise.  Would you rather I open a separate issue?

**Comment by SergioRAgostinho** [comment]
No worries and thanks for taking the time. Please do open a new one!

**Comment by SergioRAgostinho** [comment]
I just tested this PR on both Linux and Mac and it's doing what it's supposed to. All previous reports for Windows you've all given, indicate that there's still a problem. 

I thought about installing Windows on a VM to give a try. Apparently Microsoft provided Development VMs before over here https://developer.microsoft.com/en-us/windows/downloads/virtual-machines but this is no longer working. Does anyone know if there's now another way to get access to those VMs?

**Comment by taketwo** [comment]
Sorry, no idea about the images.

Since this is a blocker and in fact quite a number of people and dependent projects suffer from this, I think we should revisit my earlier suggestion: https://github.com/PointCloudLibrary/pcl/pull/2100#issuecomment-351139644.

**Comment by SergioRAgostinho** [comment]
> Since this is a blocker and in fact quite a number of people and dependent projects suffer from this, I think we should revisit my earlier suggestion: #2100 (comment).

What if we merge and we wait until windows users complain and help debug for a solution? When things work users are silent nothing changes. If there's a problem, users complain and they are more likely to help in reaching a solution. 🙄 😇

**Comment by taketwo** [comment]
I'm against taking hostages...

**Comment by SergioRAgostinho** [comment]
It's now working on Windows. :shipit:

**Comment by taketwo** [comment]
Wow, these are damn good news! So I suppose you have tested with building PCL and then configuring another project to depend upon it? If this works, then what remains is to check that @UnaNancyOwen can make a working installer with this PR, and also that vcpkg is still working?

**Comment by SergioRAgostinho** [comment]
> So I suppose you have tested with building PCL and then configuring another project to depend upon it?

Yes. I could have been more thorough in testing with one of those examples which just segfaulted out of nowhere before like #2013, but ended up testing only a single instantiation of a point type.  I tried both Debug and Release configurations.

**Comment by frozar** [comment]
That's a super good news!

**Comment by UnaNancyOwen** [comment]
I build PCL that is this pull request of latest version.
It seems correct works with traditional style (that using <code>PCL_INCLUDE_DIRS</code>, <code>PCL_DEFINITIONS</code>, <code>PCL_LIBRARIES</code>) CMakeLists on Windows. But, I can't confirm with modern style (that using <code>pcl::pcl</code>) CMakeLists yet.

**Comment by taketwo** [comment]
> But, I can't confirm with modern style (that using pcl::pcl) CMakeLists yet.

There is no such thing is this PR. What we are trying to achieve here is to make it sufficient to write

```cmake
target_link_libraries(exe, ${PCL_LIBRARIES})
```

without `PCL_INCLUDE_DIRS`, `PCL_DEFINITIONS`, etc.

**Comment by SergioRAgostinho** [comment]
>  But, I can't confirm with modern style (that using pcl::pcl) CMakeLists yet.

The modern style will need to be deferred to a later version if you people don't mind. It's important to not delay 1.9.0 any further.

**Comment by UnaNancyOwen** [comment]
@taketwo It works. Thanks.
```cmake
cmake_minimum_required(VERSION 3.9)
add_executable(project main.cpp)
find_package(PCL 1.8.1.99 REQUIRED)
target_link_libraries(project ${PCL_LIBRARIES})
```

**Comment by UnaNancyOwen** [comment]
@SergioRAgostinho I agree.

**Comment by taketwo** [comment]
Let me have another last look at this whole thing today and then, if nobody objects, merge it later this week?

**Comment by SergioRAgostinho** [comment]
Should be good now.

**Comment by taketwo** [comment]
I updated the PR description to make clear that only `target_link_libraries` is needed from now on.

**Comment by SergioRAgostinho** [comment]
We just fixed our Windows CI because of this one 🎉

**Comment by claudiofantacci** [comment]
> What if we merge and we wait until windows users complain and help debug for a solution? When things work users are silent nothing changes. If there's a problem, users complain and they are more likely to help in reaching a solution. 🙄 😇

Here we are 🤣 

I just came through the fact that with the following line
https://github.com/PointCloudLibrary/pcl/pull/2100/files#diff-af3b638bc2a3e6c650974192a53c7291R110
under Windows 10 and by using vcpkg toolchain, `CMAKE_CXX_FLAGS` is never empty and `SSE` options are never tested/enabled.

Is this still something that need to be addressed under Windows, or am I having an unexpected behaviour 😄?

**Comment by taketwo** [comment]
Which flags does _vcpkg_ pass?

The currently implemented logic is as follows: if `CMAKE_CXX_FLAGS` are not set, then add some flags that PCL thinks are necessary (including architecture and SSE flags). Conversely, if `CMAKE_CXX_FLAGS` is set explicitly by the user, then don't touch everything, assume the user knows exactly what he wants.

This logic existed before this PR. Only thing it changed is fixed a potential [nasty bug](https://github.com/PointCloudLibrary/pcl/pull/2100#issuecomment-368822323) associated with it.

In general, the fact that _vcpkg_ explicitly sets flags warrants a revision of the logic. Perhaps `CMAKE_CXX_FLAGS` should rather be treated as extra flags that are to be appended to all other flags set by PCL.

**Comment by claudiofantacci** [comment]
By default I have `/DWIN32 /D_WINDOWS /W3 /GR /EHsc` just before the `if()` statement preceding SSE macro. I'm trying to undesrtand whether this is due to `vcpkg` or by default (at least on my system).

**Comment by taketwo** [comment]
> I'm trying to undesrtand whether this is due to vcpkg or by default (at least on my system).

This would be very interesting to know indeed.

**Comment by claudiofantacci** [comment]
> > I'm trying to undesrtand whether this is due to vcpkg or by default (at least on my system).
> 
> This would be very interesting to know indeed.
> 
> **Edit:** according to [this comment](https://github.com/PointCloudLibrary/pcl/blob/a602b19cbb392bc16a26498443c3ba5010bd45f3/CMakeLists.txt#L162) these are defaults. So _vcpkg_ is not passing any additional flags. Therefore, we have a flaw in our logic.

You anticipated me 😁 
This will give a great boost to programs under Windows.
I had some doubts becuase PCL-based programs were slower under new generation Windows-based machines compared to older generation Linux/macOS-based machines.
Thanks for the PR 👍

**Pull request merged**